expan/core/experiment.py

```diff
@@ -1,280 +1,231 @@
+        :param testmethod: analysis method
+        :type  testmethod: str
+        :param **worker_args: additional arguments for the analysis method
+        :return: statistical result of the test
+        :rtype: StatisticalTestResult
+        """
+        if not isinstance(test, StatisticalTest):
+            raise TypeError("Statistical test should be of type StatisticalTest.")
+
+        if 'entity' not in self.data.columns():
+            raise RuntimeError("There is no 'entity' column in the data.")
+        if self.data.entity.duplicated().any():
+            raise ValueError('Entities in data should be unique.')
+
+        if test.variants.variant_column_name not in self.data.columns():
+            raise RuntimeError("There is no '{}' column in the data.".format(test.variants.variant_column_name))
+        if test.variants.treatment_name not in np.unique(self.data[test.variants.variant_column_name]):
+            raise RuntimeError("There is no treatment with the name '{}' in the data.".format(test.variants.treatment_name))
+        if test.variants.control_name not in np.unique(self.data[test.variants.variant_column_name]):
+            raise RuntimeError("There is no control with the name '{}' in the data.".format(test.variants.control_name))
+
+        if not isinstance(test.features, list):
+            raise TypeError("Features should be a list.")
+        if not all(isinstance(n, FeatureFilter) for n in test.features):
+            raise TypeError("Some features are not of the type FeatureFilter.")
+        for feature in test.features:
+            if feature.column_name not in self.data.columns:
+                raise RuntimeError("Feature name '{}' does not exist in the data.".format(feature.column_name))
+
+        if type(test.kpi) is KPI and (test.kpi.name not in self.data.columns()):
+            raise RuntimeError("There is no column of name '{}' in the data.".format(test.kpi.name))
+        if type(test.kpi) is DerivedKPI:
+            if type(test.kpi.formula) is not str:
+                raise RuntimeError("Formula of derived KPI '{}' does not exist.".format(test.kpi.name))
+            # create the derived kpi column if it is not yet created
+            if not test.kpi.name in self.data.columns:
+                test.kpi.make_derived_kpi(self.data)
+
+        logger.info("One analysis with kpi '{}', control variant '{}', treatment variant '{}' and features [{}] "
+                    "has just started".format(test.kpi_name, test.variants.control_name,
+                                              test.variants.treatment_name,
+                                              [(feature.column_name, feature.column_value) for feature in test.features]))
+
+        if not testmethod in self.worker_table:
+            raise NotImplementedError("Test method '{}' is not implemented.".format(testmethod))
+        worker = self.worker_table[testmethod](**worker_args)
+
+        data_for_analysis = self.data.copy()
```

Comment:
I think we don't need to make a copy here

Slicing DF later on is not destructive, but rather creates a new reference

Reply:
Good point!